docs/developer-tools/ThrowHelper.md

```diff
@@ -0,0 +1,165 @@
+    
+    public int GetValueIfNotZeroOrThrow()
+    {
+        if (number == 0)
+        {
+            throw new InvalidOperationException("The value must be != 0");
+        }
+        
+        return number;
+    }
+}
+```
+
+This type is only meant to showcase the use of `ThrowHelper` APIs. Here we're simply checking the value of the `number` field, and throwing if it's 0. Let's check the generated asm code:
+
+```x86asm
+ExceptionTest.GetValueIfNotZeroOrThrow()
+    mov eax, [rcx]              ; read number
+    test eax, eax               ; if (number != 0)
+    je short L0011              ; jump to faulting path (if == 0)
+    ret                         ; return (number is already in eax)
+    mov rcx, 0x7ff95cbaca80     ; exception setup...
+    call 0x00007ff9bc6178f0
+    mov rsi, rax
+    mov ecx, 1
+    mov rdx, 0x7ff96590c080
+    call 0x00007ff9bc7403e0
+    mov rdx, rax
+    mov rcx, rsi
+    call System.InvalidOperationException..ctor(System.String)
+    mov rcx, rsi
+    call 0x00007ff9bc5db3a0
+    int3 ; finally throw the exception
+```
+
+We can see the code contains a lot of lines just dedicated to creating the exception being thrown. This can cause a large increase in the code size whenever we have many of these checks, because it will reduce the efficiency of the instruction cache, etc. In general, we'd like the size of our methods to be as small as possible.
+
+The above snippet, rewritten using the `ThrowHelper` APIs, would like like this:
+
+```csharp
+public int GetValueIfNotZeroOrThrow()
+{
+    if (number == 0)
+    {
+        ThrowHelper.ThrowInvalidOperationException("The value must be != 0");
+    }
+
+    return number;
+}
+```
+
+Which results in the following assembly:
+
+```x86asm
+ExceptionTest.GetValueIfNotZeroOrThrow2()
+    mov eax, [rcx]          ; load number
+    test eax, eax           ; if (number != 0)
+    je short L000f          ; faulting path (if == 0)
+    ret                     ; return number (already in eax)
+    mov rcx, 0x23435d85b98  ; load the exception message
+    mov rcx, [rcx]
+    call ThrowHelper.ThrowInvalidOperationException(System.String) ; ThrowHelper call!
+    int3 ; return with fault
+```
+
+We can see that the resulting assembly is much smaller than before. We basically moved all the code to create the exception out of the method, which also means we can just always reuse the same code from the `ThrowHelper` APIs, with different arguments, instead of inlining the exception throw in every single one of our methods. Here we only have those two `mov` instructions to load the `string` with our exception type, and then we jump to the `ThrowHelper.ThrowInvalidOperationException` method. The JIT compiler is able to see that that method will always throw, so it will never inline that call, and it will make sure rewrite our conditional branches in the best way possible (specifically, knowing what is the branch that is supposed to be executed, without faulting).
```

Comment:
```suggestion
We can see that the resulting assembly is much smaller than before. We basically moved all the code to create the exception out of the method, which also means we can just always reuse the same code from the `ThrowHelper` APIs, with different arguments, instead of inlining the exception throw in every single one of our methods. Here we only have those two `mov` instructions to load the `string` with our exception type, and then we jump to the `ThrowHelper.ThrowInvalidOperationException` method. The JIT compiler is able to see that that method will always throw, so it will never inline that call, and it will make sure to rewrite our conditional branches in the best way possible (specifically, knowing what is the branch that is supposed to be executed, without faulting).
```